// @okxweb3/coin-bitcoin/dist/cat20/common/cat20Enum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MinterType = exports.CHANGE_MIN_POSTAGE = exports.Postage = exports.SupportedNetwork = exports.CatAddressType = void 0;
var CatAddressType;
(function (CatAddressType) {
    CatAddressType["P2WPKH"] = "p2wpkh";
    CatAddressType["P2TR"] = "p2tr";
})(CatAddressType = exports.CatAddressType || (exports.CatAddressType = {}));
var SupportedNetwork;
(function (SupportedNetwork) {
    SupportedNetwork["btcSignet"] = "btc-signet";
    SupportedNetwork["fractalMainnet"] = "fractal-mainnet";
    SupportedNetwork["fractalTestnet"] = "fractal-testnet";
})(SupportedNetwork = exports.SupportedNetwork || (exports.SupportedNetwork = {}));
var Postage;
(function (Postage) {
    Postage[Postage["METADATA_POSTAGE"] = 546] = "METADATA_POSTAGE";
    Postage[Postage["GUARD_POSTAGE"] = 332] = "GUARD_POSTAGE";
    Postage[Postage["MINTER_POSTAGE"] = 331] = "MINTER_POSTAGE";
    Postage[Postage["TOKEN_POSTAGE"] = 330] = "TOKEN_POSTAGE";
})(Postage = exports.Postage || (exports.Postage = {}));
exports.CHANGE_MIN_POSTAGE = 546;
var MinterType;
(function (MinterType) {
    MinterType["OPEN_MINTER_V1"] = "21cbd2e538f2b6cc40ee180e174f1e25";
    MinterType["OPEN_MINTER_V2"] = "a6c2e92d74a23c07bb6220b676c6cb9b";
    MinterType["UNKOWN_MINTER"] = "unkown_minter";
})(MinterType = exports.MinterType || (exports.MinterType = {}));
